fission-web-server/library/Fission/Web/Server/Types.hs

```diff
@@ -298,6 +301,32 @@ instance MonadDNSLink Server where
 instance MonadLinkedIPFS Server where
   getLinkedPeers = asks ipfsRemotePeers
 
+instance MonadIPFSPinner Server where
+  pin cid = 
+    asks clusterURL >>= \case
+      Nothing -> 
+        IPFS.Pin.add cid >>= \case
+          Left err -> 
+            return $ Error.openLeft err
+          Right _ -> 
+            return $ Right ()
+
+      Just (IPFS.URL url) -> do
+        manager      <- asks httpManager
+        let 
+          query = Cluster.pinClient cid
+          env = mkClientEnv manager url
+        
+        (liftIO $ runClientM query env) >>= \case
+          Left err -> do
+            formattedErr <- Cluster.parseClientError err
+            return $ Error.openLeft formattedErr
+
+          Right _ -> 
+            return $ Right ()
```

Comment:
Totally fine as-is, but if you want to save a couple keystrokes:

```haskell
liftIO (runClientM query env) >>= \case
  -- ...
  Right _ -> 
    return ok
```

Reply:
ah yea nice! forgot about `ok`